src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs

```diff
@@ -346,83 +342,78 @@ public Assembly LoadFromPath(string originalAnalyzerPath)
+        /// <param name="windowsShadowPath">A shadow copy path will be created on Windows and this value 
+        /// will be the base directory where shadow copy assemblies are stored. </param>
+        internal static IAnalyzerAssemblyLoaderInternal CreateNonLockingLoader(
+            string windowsShadowPath,
+            ImmutableArray<IAnalyzerPathResolver> pathResolvers = default,
+            ImmutableArray<IAnalyzerAssemblyResolver> assemblyResolvers = default,
+            System.Runtime.Loader.AssemblyLoadContext? compilerLoadContext = null)
         {
-            CheckIfDisposed();
+            pathResolvers = pathResolvers.NullToEmpty();
+            assemblyResolvers = assemblyResolvers.NullToEmpty();
 
-            lock (_guard)
+            if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                return new AnalyzerAssemblyLoader(
+                    pathResolvers,
+                    [.. assemblyResolvers, StreamResolver.Instance],
+                    compilerLoadContext);
+            }
+
+            // The shadow copy analyzer should only be created on Windows. To create on Linux we cannot use 
+            // GetTempPath as it's not per-user. Generally there is no need as LoadFromStream achieves the same
+            // effect
+            if (!Path.IsPathRooted(windowsShadowPath))
             {
-                return _analyzerAssemblyInfoMap
-                    .Select(x => (x.Key, x.Value?.RealAssemblyPath ?? ""))
-                    .OrderBy(x => x.Key)
-                    .ToArray();
+                throw new ArgumentException("Must be a full path.", nameof(windowsShadowPath));
             }
+
+            return new AnalyzerAssemblyLoader(
+                [.. pathResolvers, new ProgramFilesAnalyzerPathResolver(), new ShadowCopyAnalyzerPathResolver(windowsShadowPath)],
+                [.. assemblyResolvers, DiskResolver.Instance],
+                compilerLoadContext);
         }
 
+#else
+
         /// <summary>
-        /// Iterates the <see cref="_externalResolvers"/> if any, to see if any of them can resolve
-        /// the given <see cref="AssemblyName"/> to an <see cref="Assembly"/>.
+        /// Return an <see cref="IAnalyzerAssemblyLoader"/> which does not lock assemblies on disk that is
+        /// most appropriate for the current platform.
         /// </summary>
-        /// <param name="assemblyName">The name of the assembly to resolve</param>
-        /// <returns>An <see langword="assembly"/> if one of the resolvers is successful, or <see langword="null"/></returns>
-        internal Assembly? ResolveAssemblyExternally(AssemblyName assemblyName, string rootDirectory)
+        /// <param name="windowsShadowPath">A shadow copy path will be created on Windows and this value 
+        /// will be the base directory where shadow copy assemblies are stored. </param>
+        internal static IAnalyzerAssemblyLoaderInternal CreateNonLockingLoader(
```

Comment:
slightly surprised/curious about neither of the two CreateNonLockingLoader methods calling into each other.

Reply:
Two of the three parameters are .NET Core only types and that makes this challenging to unify. I had one version of the change where there was a method they both called into and I found it less readable than the current version.